---
title: "Lab 11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## A class constructor

Let's make an S3 class that will contain data about a single gene.  It will 
contain the gene ID, gene symbol, a vector of ontology terms, and the 
protein-coding portion of the DNA sequence (the CDS).  Below is a function
to serve as the class constructor.

```{r error = TRUE, echo = TRUE}
# Constructor function for the 'Gene' class
Gene <- function(ID, symbol, ontology, CDS){
  if(length(ID) > 1) stop("Only provide one ID")
  if(length(symbol) > 1) stop("Only provide one symbol")
  if(length(CDS) > 1) stop("Only provide one CDS")
  if(!is.character(ID)) stop("ID must be a character string")
  if(!is.character(symbol)) stop("symbol must be a character string")
  if(!is.numeric(ontology)) stop("ontology must be a vector of GO numbers")
  if(!is.character(CDS)) stop("CDS must be a character string")
  
  # clean out line breaks from CDS and check that it is DNA sequence
  CDS <- gsub("\n", "", CDS)
  if(nchar(CDS) %% 3 != 0){
    stop("Length of protein coding sequence must be a multiple of three.")
  }
  if(!all(strsplit(CDS, "")[[1]] %in% c("A", "C", "G", "T"))){
    stop("CDS must be DNA sequence.")
  }
  
  out <- list(ID = ID, symbol = symbol, ontology = as.integer(ontology),
              CDS = CDS)
  return(out)
}

## Test out the constructor
phyA <- Gene("Solyc10g044670.1.1", "Phytochrome A", 
             c(155, 5515, 6355, 7165, 9584, 18298),
             "ATGTCGTCTTCAAGACCTAGCCAATCTTCCACCACTTCATCAAGATCAAAGCACAGTGCTAGGATCGTTGCACAGACCTCTATAGATGCAAAGCTGCATG
CAGACTTTGAGGAGTCAGGTGATTCCTTTGACTATTCAAGCTCAGTGAGGGTTACAAGTGTCGCTGGAGATGAGGAAAAACCAAAGTCAGACAAAGTTAC
CACCGCCTACCTTCATCAGATCCAAAAGGGCAAGTTTATCCAGCCATTTGGTTGTCTGTTAGCCCTGGATGAGAAAACATTAAAGGTCATAGCATTCAGT
GAGAATGCCCCTGAAATGCTGACCATGGTTAGCCATGCTGTTCCAAGTGTTGGTGAGCATCCAGTTCTTGGCATCGGGACTGATATCAGAACGATCTTCA
CTGGTCCTAGTGGCGCAGCATTGCAGAAAGCCTTGGGGTTTGGGGAGGTTTCTCTGTTAAATCCTGTCCTTGTTCACTGTAAAAATTCTGGAAAGCCATT
TTATGCAATTGTTCATAGGGTTACAGGTAGCTTAATCCTTGATTTTGAGCCTGTGAAGCCCTATGAAGTACCCATGACTGCTGCAGGGGCCCTGCAGTCA
TATAAACTAGCAGCCAAAGCCATTACTCGCTTGCAGTCCTTGCCCAGTGGCAGTATGGAAAGACTCTGTGACACAATGGTTCAGGAGGTTTTTGAACTCA
CAGGTTATGACAGGGTGATGGGATATAAGTTTCACGAGGATGATCATGGAGAGGTGGTGTCTGAGATCACAAAGCCTGGCCTCGAGCCTTACCTTGGTTT
ACATTATCCTGCTACAGATATTCCACAGGCTGCACGGTTTTTGTTTATGAAGAATAAGGTCCGAATGATTTGTGATTGCCGAGCAAAGCATGTGAAGGTA
GTCCAAGATGAGAAGCTTCCATTTGATTTAACATTGTGCGGCTCTACTCTTAGGGCCCCTCACTACTGCCATTTACAGTATATGGAGAACATGAATTCAA
TTGCATCACTTGTAATGGCAGTTGTGGTCAATGACGGGGATGAAGAAGGAGAAAGCTCTGATTCTTCACAATCTCAAAAAAGAAAAAGGCTATGGGGCCT
GGTTGTTTGCCACAACACGACCCCAAGGTTCGTTCCCTTTCCACTGAGGTATGCATGTGAGTTTCTTGCACAAGTCTTTGCCATACACGTCAACAAGGAA
CTGGAATTGGAAAATCAATTCCTTGAGAAAAATATTCTGCGTACTCAGACTCTCTTGTGTGATATGCTGATGCGAGATGCTCCCCTAGGTATCGTGTCAC
AGAGCCCCAACATTATGGATCTTGTCAAATGTGATGGTGCGGCTTTGCTCTATAAGAATAAGATACATCGATTAGGAATGAACCCAAGTGACTTTCAGCT
GCAGGATATAGTATCATGGCTTTGTGAGTATCATACAGATTCCACAGGCTTGAGTACGGATAGCTTGTATGATGCTGGTTTTCCTGGGGCTCTTGCTCTT
GGTGATGCAGTCTGTGGTATGGCAGCTGTGAGAATATCTGATAAGGACTGGCTGTTCTGGTTCAGGTCACACACTGCTGCTGAAGTTAGATGGGGTGGTG
CAAAGCATGAACCTGGTGAGAAGGATGATGGCAGGAAAATGCATCCTAGGTCATCATTCAAAGCATTCCTGGAAGTTGTCAAGACAAGGAGTATACCCTG
GAAGGATTATGAGATGGATGCAATCCACTCTTTGCAGCTCATACTAAGAAATGCTTTCAAGGATGCTGAGGTTGTGAATTCAAATACCAATTCCATCTAT
AAGAAGCTTAATGATCTAAAGATTGACGGAATGCAGGAACTAGAATCAGTGACTGCTGAAATGGTCCGCTTAATTGAAACAGCTTTAGTTCCTATATTAG
CAGTTGATGTTGATGGGCAGGTTAATGGATGGAACACAAAAATAGCTGAATTAACTGGTCTTCCTGTTGATGAAGCAATTGGGAAACATCTTCTCACTCT
CGTGGAGGATTCATCAGTTGATACCGTGAATAAGATGTTGGAATTAGCATTGCAGGGGAAAGAGGAAAAAAATGTAGAGTTTGAAATAAAAACACATGGG
CCGTCAAGAGATTCTAGTCCAATCAGCTTAATCGTGAATGCATGTGCAAGCAAAGATGTTCGGGATAATGTTGTGGGTGTGTGTTTTATGGCTCATGATA
TAACTGGACAAAAAAGCATCATGGACAAGTTCACCCGAATCGAAGGTGACTATAGAGCGATTATCCAAAATCCTCACCCATTGATCCCACCAATATTTGG
CACTGATCAATTTGGCTGGTGTTCTGAGTGGAACACTGCAATGACCAAGTTAACTGGATGGCGGCGTGATGATGTTATGGATAAAATGCTGTTAGGGGAG
GTTTTTGGGACACAGGCAGCTTGCTGCCGTCTCAAGAATCAAGAAGCGTTTGTAAATTTCGGAGTTGTACTAAACAATGCTATCACTGGTCAAGAATCTG
AGAAGATTCCTTTTGGTTTCTTTGCACGTTACGGGAAATATGTGGAGTGCTTACTCTGTGTGAGCAAAAGGTTAGATAAAGAGGGTGCAGTCACGGGACT
CTTTTGTTTCCTGCAACTTGCAAGCCATGAGTTGCAACAAGCTCTTTATGTTCAACGATTATCAGAACAAACTGCGTTGAAAAGATTGAAAGTATTAGCT
TACATAAGGAGGCAAATTAGGAATCCTCTTTCTGGGATTATATTCTCTCGGAAGATGCTAGAGGGGACTAGCTTGGGCGAAGAGCAGAAAAATATACTGC
ATACAAGTGCACAGTGTCAGCGTCAGCTCAACAAAATTCTTGATGATACAGATCTTGATAGCATCATAGATGGTTATTTGGATCTGGAGATGCTCGAGTT
TAAGCTGCATGAAGTATTGGTAGCATCTATAAGTCAAGTCATGATGAAGAGCAACGGAAAGAATATAATGATCTCGAATGACATGGTTGAAGATCTTCTT
AATGAAACTTTATACGGAGATAGTCCGAGGCTTCAACAAGTCCTTGCTAACTTTTTGTTAGTATCTGTGAACGCTACACCAAGTGGCGGCCAGCTTAGTA
TTTCAGGCAGGCTAACAAAAGATCGTATAGGAGAATCTGTTCAGCTTGCTCTCTTGGAATTCAGGATAAGACATACAGGGGGTGGAGTGCCAGAAGAATT
GCTTGGCCAAATGTTTGGTAGTGAAGCCGATGCATCTGAAGAAGGGATCAGCTTACTTGTCAGCAGAAAACTGGTCAAGCTGATGAATGGGGAAGTTCAG
TATCTAAGAGAGGCTGGGCAATCAACTTTCATTATATCCGTTGAACTCGCAGTGGCTACCAACTCAAGCTGA")

phyA

# test one of the errors
phyA <- Gene("Solyc10g044670.1.1", "Phytochrome A", 
             c(155, 5515, 6355, 7165, 9584, 18298),
             "MXXRRGATBB")
```

**Question 1 to turn in (1 point):**  If we want to be able to create methods for the 
`Gene` class, we need to add one more line to the constructor function.  What 
is that line?

```{r}
# The line is:
# class(out) <- c("Gene", class(out))

# Contstructor function for the 'Gene' class
Gene <- function(ID, symbol, ontology, CDS){
  if(length(ID) > 1) stop("Only provide one ID")
  if(length(symbol) > 1) stop("Only provide one symbol")
  if(length(CDS) > 1) stop("Only provide one CDS")
  if(!is.character(ID)) stop("ID must be a character string")
  if(!is.character(symbol)) stop("symbol must be a character string")
  if(!is.numeric(ontology)) stop("ontology must be a vector of GO numbers")
  if(!is.character(CDS)) stop("CDS must be a character string")
  
  # clean out line breaks from CDS and check that it is DNA sequence
  CDS <- gsub("\n", "", CDS)
  if(nchar(CDS) %% 3 != 0){
    stop("Length of protein coding sequence must be a multiple of three.")
  }
  if(!all(strsplit(CDS, "")[[1]] %in% c("A", "C", "G", "T"))){
    stop("CDS must be DNA sequence.")
  }
  
  out <- list(ID = ID, symbol = symbol, ontology = as.integer(ontology),
              CDS = CDS)
  class(out) <- c("Gene", class(out))
  return(out)
}

phyA <- Gene("Solyc10g044670.1.1", "Phytochrome A", 
             c(155, 5515, 6355, 7165, 9584, 18298),
             "ATGTCGTCTTCAAGACCTAGCCAATCTTCCACCACTTCATCAAGATCAAAGCACAGTGCTAGGATCGTTGCACAGACCTCTATAGATGCAAAGCTGCATG
CAGACTTTGAGGAGTCAGGTGATTCCTTTGACTATTCAAGCTCAGTGAGGGTTACAAGTGTCGCTGGAGATGAGGAAAAACCAAAGTCAGACAAAGTTAC
CACCGCCTACCTTCATCAGATCCAAAAGGGCAAGTTTATCCAGCCATTTGGTTGTCTGTTAGCCCTGGATGAGAAAACATTAAAGGTCATAGCATTCAGT
GAGAATGCCCCTGAAATGCTGACCATGGTTAGCCATGCTGTTCCAAGTGTTGGTGAGCATCCAGTTCTTGGCATCGGGACTGATATCAGAACGATCTTCA
CTGGTCCTAGTGGCGCAGCATTGCAGAAAGCCTTGGGGTTTGGGGAGGTTTCTCTGTTAAATCCTGTCCTTGTTCACTGTAAAAATTCTGGAAAGCCATT
TTATGCAATTGTTCATAGGGTTACAGGTAGCTTAATCCTTGATTTTGAGCCTGTGAAGCCCTATGAAGTACCCATGACTGCTGCAGGGGCCCTGCAGTCA
TATAAACTAGCAGCCAAAGCCATTACTCGCTTGCAGTCCTTGCCCAGTGGCAGTATGGAAAGACTCTGTGACACAATGGTTCAGGAGGTTTTTGAACTCA
CAGGTTATGACAGGGTGATGGGATATAAGTTTCACGAGGATGATCATGGAGAGGTGGTGTCTGAGATCACAAAGCCTGGCCTCGAGCCTTACCTTGGTTT
ACATTATCCTGCTACAGATATTCCACAGGCTGCACGGTTTTTGTTTATGAAGAATAAGGTCCGAATGATTTGTGATTGCCGAGCAAAGCATGTGAAGGTA
GTCCAAGATGAGAAGCTTCCATTTGATTTAACATTGTGCGGCTCTACTCTTAGGGCCCCTCACTACTGCCATTTACAGTATATGGAGAACATGAATTCAA
TTGCATCACTTGTAATGGCAGTTGTGGTCAATGACGGGGATGAAGAAGGAGAAAGCTCTGATTCTTCACAATCTCAAAAAAGAAAAAGGCTATGGGGCCT
GGTTGTTTGCCACAACACGACCCCAAGGTTCGTTCCCTTTCCACTGAGGTATGCATGTGAGTTTCTTGCACAAGTCTTTGCCATACACGTCAACAAGGAA
CTGGAATTGGAAAATCAATTCCTTGAGAAAAATATTCTGCGTACTCAGACTCTCTTGTGTGATATGCTGATGCGAGATGCTCCCCTAGGTATCGTGTCAC
AGAGCCCCAACATTATGGATCTTGTCAAATGTGATGGTGCGGCTTTGCTCTATAAGAATAAGATACATCGATTAGGAATGAACCCAAGTGACTTTCAGCT
GCAGGATATAGTATCATGGCTTTGTGAGTATCATACAGATTCCACAGGCTTGAGTACGGATAGCTTGTATGATGCTGGTTTTCCTGGGGCTCTTGCTCTT
GGTGATGCAGTCTGTGGTATGGCAGCTGTGAGAATATCTGATAAGGACTGGCTGTTCTGGTTCAGGTCACACACTGCTGCTGAAGTTAGATGGGGTGGTG
CAAAGCATGAACCTGGTGAGAAGGATGATGGCAGGAAAATGCATCCTAGGTCATCATTCAAAGCATTCCTGGAAGTTGTCAAGACAAGGAGTATACCCTG
GAAGGATTATGAGATGGATGCAATCCACTCTTTGCAGCTCATACTAAGAAATGCTTTCAAGGATGCTGAGGTTGTGAATTCAAATACCAATTCCATCTAT
AAGAAGCTTAATGATCTAAAGATTGACGGAATGCAGGAACTAGAATCAGTGACTGCTGAAATGGTCCGCTTAATTGAAACAGCTTTAGTTCCTATATTAG
CAGTTGATGTTGATGGGCAGGTTAATGGATGGAACACAAAAATAGCTGAATTAACTGGTCTTCCTGTTGATGAAGCAATTGGGAAACATCTTCTCACTCT
CGTGGAGGATTCATCAGTTGATACCGTGAATAAGATGTTGGAATTAGCATTGCAGGGGAAAGAGGAAAAAAATGTAGAGTTTGAAATAAAAACACATGGG
CCGTCAAGAGATTCTAGTCCAATCAGCTTAATCGTGAATGCATGTGCAAGCAAAGATGTTCGGGATAATGTTGTGGGTGTGTGTTTTATGGCTCATGATA
TAACTGGACAAAAAAGCATCATGGACAAGTTCACCCGAATCGAAGGTGACTATAGAGCGATTATCCAAAATCCTCACCCATTGATCCCACCAATATTTGG
CACTGATCAATTTGGCTGGTGTTCTGAGTGGAACACTGCAATGACCAAGTTAACTGGATGGCGGCGTGATGATGTTATGGATAAAATGCTGTTAGGGGAG
GTTTTTGGGACACAGGCAGCTTGCTGCCGTCTCAAGAATCAAGAAGCGTTTGTAAATTTCGGAGTTGTACTAAACAATGCTATCACTGGTCAAGAATCTG
AGAAGATTCCTTTTGGTTTCTTTGCACGTTACGGGAAATATGTGGAGTGCTTACTCTGTGTGAGCAAAAGGTTAGATAAAGAGGGTGCAGTCACGGGACT
CTTTTGTTTCCTGCAACTTGCAAGCCATGAGTTGCAACAAGCTCTTTATGTTCAACGATTATCAGAACAAACTGCGTTGAAAAGATTGAAAGTATTAGCT
TACATAAGGAGGCAAATTAGGAATCCTCTTTCTGGGATTATATTCTCTCGGAAGATGCTAGAGGGGACTAGCTTGGGCGAAGAGCAGAAAAATATACTGC
ATACAAGTGCACAGTGTCAGCGTCAGCTCAACAAAATTCTTGATGATACAGATCTTGATAGCATCATAGATGGTTATTTGGATCTGGAGATGCTCGAGTT
TAAGCTGCATGAAGTATTGGTAGCATCTATAAGTCAAGTCATGATGAAGAGCAACGGAAAGAATATAATGATCTCGAATGACATGGTTGAAGATCTTCTT
AATGAAACTTTATACGGAGATAGTCCGAGGCTTCAACAAGTCCTTGCTAACTTTTTGTTAGTATCTGTGAACGCTACACCAAGTGGCGGCCAGCTTAGTA
TTTCAGGCAGGCTAACAAAAGATCGTATAGGAGAATCTGTTCAGCTTGCTCTCTTGGAATTCAGGATAAGACATACAGGGGGTGGAGTGCCAGAAGAATT
GCTTGGCCAAATGTTTGGTAGTGAAGCCGATGCATCTGAAGAAGGGATCAGCTTACTTGTCAGCAGAAAACTGGTCAAGCTGATGAATGGGGAAGTTCAG
TATCTAAGAGAGGCTGGGCAATCAACTTTCATTATATCCGTTGAACTCGCAGTGGCTACCAACTCAAGCTGA")
```

## Methods for the 'Gene' class

Now that we have the class set up, we can create some methods for it.

**Question 2 to turn in (1 point):**  Make a generic function called `proteinLength` with a 
method for `Gene`, which gives the length, in amino acids, of the protein corresponding to
the CDS in the `Gene` object.  (This is the length of the CDS in nucleotides divided by three.)  Also make
a `print` method for `Gene` that uses the `cat` function to display some information about
the object.

```{r}
proteinLength <- function(object, ...){
  UseMethod("proteinLength", object)
}
proteinLength.Gene <- function(object, ...){
  return(nchar(object$CDS) / 3L)
}

print.Gene <- function(object){
  cat(paste("Gene ID:", object$ID),
      paste("Symbol:", object$symbol),
      paste(proteinLength(object), "amino acids"),
      sep = "\n")
}
```

```{r echo = TRUE}
proteinLength(phyA)

phyA
```

## Incorporating classes and methods into an R Package

Fork the repository at https://github.com/lvclark/Lab11Gene and clone
your forked version to your computer to edit it.

In the R folder, there is a file called classes_and_methods.R.  Edit
the constructor function the way that you did for question 1.  Add in
your function and method definitions from question 2.

Make one or two additional .Rd files describing your methods from question 2 and put them
in the man folder.  Be sure to have aliases both for the `proteinLength` generic
function, and for the `print.Gene` and `proteinLength.Gene` methods.  You'll also
need usage sections for all three.  The usage section will be in the format:

```
\usage{
\method{function}{class}(arguments, ...)
}
```

Make sure your package builds and passes the check.

**Question 3 to turn in (1 point):** Turn in the URL to your forked respository.

(You don't need to do a pull request.)

## Other stuff to think about

If we do 

```{r echo = TRUE}
class(phyA)
```

we see that `phyA` is still a list in addition to being a `Gene` object.  Why do we want 
to have it still be a `list`, and not have `Gene` be the only class?  Why do we put the
classes in this order?

```{r}
## By leaving it as a list, we can still use list methods on the object, like
## `sapply` or list indexing.  We have `Gene` first though so that R will first
## look for `Gene` methods, then only use `list` methods if it can't find a
## `Gene` method.
```

Think about what we learned last week about optimization in R.  If you were
processing thousands of genes, you could store them as a list of `Gene` 
objects.  Would this be a good approach?

```{r}
## Probably not, since you would have to do all processing on the gene list in 
## a loop or using `lapply` or `sapply`, which would slow things down.  You
## might instead want a class to contain all the genes, with a vector of 
## gene IDs, a vector of CDS sequences, etc.  Next week we'll start learning
## Bioconductor, which has these sorts of classes.
##
## Note that in most other languages, such as Python, C++, etc., there 
## would be nothing wrong with making a list of Gene objects and looping
## through them.
```

Make some accessor functions for the slots of `Gene` objects.  Can you make
a function that returns the ontology terms in the format "GO:0000115" (even 
though that is not the format they are in inside the object)?

```{r}
ID <- function(object, ...){
  UseMethod("ID", object)
}
ID.Gene <- function(object, ...){
  return(object$ID)
}
symbol <- function(object, ...){
  UseMethod("symbol", object)
}
symbol.Gene <- function(object, ...){
  return(object$symbol)
}
CDS <- function(object, ...){
  UseMethod("CDS", object)
}
CDS.Gene <- function(object, ...){
  return(object$CDS)
}
ontology <- function(object, ...){
  UseMethod("ontology", object)
}
ontology.Gene <- function(object, ...){
  gostrings <- paste("GO:",
                     formatC(object$ontology, flag = "0", width = 7),
                     sep = "")
  return(gostrings)
}
```

```{r echo = TRUE}
ID(phyA)
ontology(phyA)
```

